torch/csrc/jit/passes/onnx/fold_if_node.cpp

```diff
@@ -0,0 +1,229 @@
+#include <torch/csrc/jit/passes/onnx/fold_if_node.h>
+#include <torch/csrc/jit/passes/onnx/helper.h>
+#include <torch/csrc/jit/passes/onnx/scalar_type_analysis.h>
+#include <torch/torch.h>
+
+#include <c10/util/Optional.h>
+#include <algorithm>
+
+namespace torch {
+namespace jit {
+
+namespace onnx {
+using namespace ::c10::onnx;
+}
+
+static bool conditionIfFold(Node* node) {
+  auto cast_node = node->input()->node();
+  if (cast_node->kind() != onnx::Cast)
+    cast_node = node;
+  auto prev_node = cast_node->input()->node();
+
+  if (prev_node->kind() == onnx::Not || prev_node->kind() == onnx::Identity ||
+      prev_node->kind() == onnx::If)
+    return conditionIfFold(prev_node);
+
+  auto compare_node = prev_node;
+  if (compare_node->kind() == onnx::Equal ||
+      compare_node->kind() == onnx::Greater ||
+      compare_node->kind() == onnx::Less) {
+    for (size_t i = 0; i < compare_node->inputs().size(); i++) {
+      auto sym = compare_node->inputs()[i]
+                     ->type()
+                     ->cast<TensorType>()
+                     ->symbolic_sizes();
+      if (!(compare_node->inputs()[i]->node()->kind() == onnx::Constant ||
+            compare_node->inputs()[i]->node()->kind() == onnx::Size ||
+            compare_node->inputs()[i]->node()->kind() == onnx::ReduceProd))
+        return false;
+      if (compare_node->inputs()[i]->node()->kind() != onnx::Constant) {
+        auto shape_node = compare_node->inputs()[i]->node()->input()->node();
+        auto shape =
+            shape_node->input()->type()->cast<TensorType>()->symbolic_sizes();
+
+        // ONNX shape and type inference cannot determine the shape of the input
+        if (!shape.rank())
+          return false;
+
+        // If dynamic_axes are used on inputs to ReduceProd node, don't fold If
+        // node
+        auto dynamic_axes = shape.isComplete();
+        if (!dynamic_axes &&
+            compare_node->inputs()[i]->node()->kind() == onnx::ReduceProd)
+          return false;
+      }
+    }
+    return true;
+  } else if (compare_node->kind() == onnx::Constant) {
+    return true;
+  }
+  return false;
+}
+
+static bool constantFoldingValue(Node* node) {
```

Comment:
Is this function also exclusively for if node? sees a lot of assumptions on input relations.